File: leetcode/c++/LCR_134_shu_zhi_de_zheng_shu_ci_fang_lcof.cpp
// https://leetcode.cn/problems/shu-zhi-de-zheng-shu-ci-fang-lcof

#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    double myPow(double x, int n) {
        long i = n;
        if (n < 0) {
            x = 1 / x;
            i = -i;
        }
        double res = 1;
        while (i > 0) {
            if (i & 1)
                res *= x;
            x *= x;
            i >>= 1;
        }
        return res;
    }
};


// 快速幂，时间复杂度O(logn)，空间复杂度为O(1)

File: leetcode/golang/LCR_134_shu_zhi_de_zheng_shu_ci_fang_lcof.go
func myPow(x float64, n int) float64 {
    if n == 0 {
        return 1
    }
    if n == -1 {
        return 1 / x
    }
    if n & 1 == 1 {
        return x * myPow(x*x, n>>1)
    }
    return myPow(x*x, n>>1)
}